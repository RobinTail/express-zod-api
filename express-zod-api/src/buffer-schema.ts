import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
