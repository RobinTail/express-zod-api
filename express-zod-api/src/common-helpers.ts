import { Request } from "express";
import * as R from "ramda";
import { z } from "zod";
import { CommonConfig, InputSource, InputSources } from "./config-type";
import { contentTypes } from "./content-type";
import { OutputValidationError } from "./errors";
import { metaSymbol } from "./metadata";
import { AuxMethod, Method } from "./method";

/** @desc this type does not allow props assignment, but it works for reading them when merged with another interface */
export type EmptyObject = Record<string, never>;
export type EmptySchema = z.ZodObject<EmptyObject, "strip">;
export type FlatObject = Record<string, unknown>;

/** @link https://stackoverflow.com/a/65492934 */
type NoNever<T, F> = [T] extends [never] ? F : T;

/**
 * @desc Using module augmentation approach you can specify tags as the keys of this interface
 * @example declare module "express-zod-api" { interface TagOverrides { users: unknown } }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- augmentation
export interface TagOverrides {}
export type Tag = NoNever<keyof TagOverrides, string>;

/** @see https://expressjs.com/en/guide/routing.html */
export const routePathParamsRegex = /:([A-Za-z0-9_]+)/g;
export const getRoutePathParams = (path: string): string[] =>
  path.match(routePathParamsRegex)?.map((param) => param.slice(1)) || [];

const areFilesAvailable = (request: Request): boolean => {
  const contentType = request.header("content-type") || "";
  const isUpload = contentType.toLowerCase().startsWith(contentTypes.upload);
  return "files" in request && isUpload;
};

export const defaultInputSources: InputSources = {
  get: ["query", "params"],
  post: ["body", "params", "files"],
  put: ["body", "params"],
  patch: ["body", "params"],
  delete: ["query", "params"],
};
const fallbackInputSource: InputSource[] = ["body", "query", "params"];

export const getActualMethod = (request: Request) =>
  request.method.toLowerCase() as Method | AuxMethod;

export const getInput = (
  req: Request,
  userDefined: CommonConfig["inputSources"] = {},
): FlatObject => {
  const method = getActualMethod(req);
  if (method === "options") return {};
  return (
    userDefined[method] ||
    defaultInputSources[method] ||
    fallbackInputSource
  )
    .filter((src) => (src === "files" ? areFilesAvailable(req) : true))
    .reduce<FlatObject>((agg, src) => Object.assign(agg, req[src]), {});
};

export const ensureError = (subject: unknown): Error =>
  subject instanceof Error ? subject : new Error(String(subject));

export const getMessageFromError = (error: Error): string => {
  if (error instanceof z.ZodError) {
    return error.issues
      .map(({ path, message }) =>
        (path.length ? [path.join("/")] : []).concat(message).join(": "),
      )
      .join("; ");
  }
  if (error instanceof OutputValidationError) {
    const hasFirstField = error.cause.issues[0]?.path.length > 0;
    return `output${hasFirstField ? "/" : ": "}${error.message}`;
  }
  return error.message;
};

/** Takes the original unvalidated examples from the properties of ZodObject schema shape */
export const pullExampleProps = <T extends z.SomeZodObject>(subject: T) =>
  Object.entries(subject.shape).reduce<Partial<z.input<T>>[]>(
    (acc, [key, schema]) => {
      const { _def } = schema as z.ZodType;
      return combinations(
        acc,
        (_def[metaSymbol]?.examples || []).map(R.objOf(key)),
        ([left, right]) => ({ ...left, ...right }),
      );
    },
    [],
  );

export const getExamples = <
  T extends z.ZodType,
  V extends "original" | "parsed" | undefined,
>({
  schema,
  variant = "original",
  validate = variant === "parsed",
  pullProps = false,
}: {
  schema: T;
  /**
   * @desc examples variant: original or parsed
   * @example "parsed" — for the case when possible schema transformations should be applied
   * @default "original"
   * @override validate: variant "parsed" activates validation as well
   * */
  variant?: V;
  /**
   * @desc filters out the examples that do not match the schema
   * @default variant === "parsed"
   * */
  validate?: boolean;
  /**
   * @desc should pull examples from properties — applicable to ZodObject only
   * @default false
   * */
  pullProps?: boolean;
}): ReadonlyArray<V extends "parsed" ? z.output<T> : z.input<T>> => {
  let examples = schema._def[metaSymbol]?.examples || [];
  if (!examples.length && pullProps && schema instanceof z.ZodObject)
    examples = pullExampleProps(schema);
  if (!validate && variant === "original") return examples;
  const result: Array<z.input<T> | z.output<T>> = [];
  for (const example of examples) {
    const parsedExample = schema.safeParse(example);
    if (parsedExample.success)
      result.push(variant === "parsed" ? parsedExample.data : example);
  }
  return result;
};

export const combinations = <T>(
  a: T[],
  b: T[],
  merge: (pair: [T, T]) => T,
): T[] => (a.length && b.length ? R.xprod(a, b).map(merge) : a.concat(b));

/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const makeCleanId = (...args: string[]) => {
  const byAlpha = R.chain((entry) => entry.split(/[^A-Z0-9]/gi), args);
  const byWord = R.chain(
    (entry) =>
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    byAlpha,
  );
  return byWord.map(ucFirst).join("");
};

export const getTransformedType = R.tryCatch(
  <T>(schema: z.ZodEffects<z.ZodTypeAny, unknown, T>, sample: T) =>
    typeof schema.parse(sample),
  R.always(undefined),
);

/** @desc can still be an array, use Array.isArray() or rather R.type() to exclude that case */
export const isObject = (subject: unknown) =>
  typeof subject === "object" && subject !== null;

export const isProduction = R.memoizeWith(
  () => process.env.TSUP_STATIC as string, // eslint-disable-line no-restricted-syntax -- substituted by TSUP
  () => process.env.NODE_ENV === "production", // eslint-disable-line no-restricted-syntax -- memoized
);
