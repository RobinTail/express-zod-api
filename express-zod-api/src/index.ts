import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
