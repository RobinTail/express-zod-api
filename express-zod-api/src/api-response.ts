import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
