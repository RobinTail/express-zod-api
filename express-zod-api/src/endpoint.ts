import { Request, Response } from "express";
import { z } from "zod";
import { NormalizedResponse, ResponseVariant } from "./api-response";
import { hasRaw, hasUpload } from "./deep-checks";
import {
  FlatObject,
  getActualMethod,
  getInput,
  ensureError,
} from "./common-helpers";
import { CommonConfig } from "./config-type";
import {
  InputValidationError,
  OutputValidationError,
  ResultHandlerError,
} from "./errors";
import { IOSchema } from "./io-schema";
import { lastResortHandler } from "./last-resort";
import { ActualLogger } from "./logger-helpers";
import { LogicalContainer } from "./logical-container";
import { AuxMethod, Method } from "./method";
import { AbstractMiddleware, ExpressMiddleware } from "./middleware";
import { ContentType } from "./content-type";
import { Nestable } from "./routable";
import { AbstractResultHandler } from "./result-handler";
import { Security } from "./security";

export type Handler<IN, OUT, OPT> = (params: {
  input: IN;
  options: OPT;
  logger: ActualLogger;
}) => Promise<OUT>;

type DescriptionVariant = "short" | "long";
type IOVariant = "input" | "output";

export abstract class AbstractEndpoint extends Nestable {
  public abstract execute(params: {
    request: Request;
    response: Response;
    logger: ActualLogger;
    config: CommonConfig;
  }): Promise<void>;
  public abstract getDescription(
    variant: DescriptionVariant,
  ): string | undefined;
  public abstract getMethods(): ReadonlyArray<Method> | undefined;
  public abstract getSchema(variant: IOVariant): IOSchema;
  public abstract getResponses(
    variant: ResponseVariant,
  ): ReadonlyArray<NormalizedResponse>;
  // @todo should return ReadonlyArray
  public abstract getSecurity(): LogicalContainer<Security>[];
  public abstract getScopes(): ReadonlyArray<string>;
  public abstract getTags(): ReadonlyArray<string>;
  public abstract getOperationId(method: Method): string | undefined;
  public abstract getRequestType(): ContentType;
}

export class Endpoint<
  IN extends IOSchema,
  OUT extends IOSchema,
  OPT extends FlatObject,
> extends AbstractEndpoint {
  readonly #descriptions: Record<DescriptionVariant, string | undefined>;
  readonly #methods?: ReadonlyArray<Method>;
  readonly #middlewares: AbstractMiddleware[];
  readonly #responses: Record<
    ResponseVariant,
    ReadonlyArray<NormalizedResponse>
  >;
  readonly #handler: Handler<z.output<IN>, z.input<OUT>, OPT>;
  readonly #resultHandler: AbstractResultHandler;
  readonly #schemas: { input: IN; output: OUT };
  readonly #scopes: ReadonlyArray<string>;
  readonly #tags: ReadonlyArray<string>;
  readonly #getOperationId: (method: Method) => string | undefined;
  readonly #requestType: ContentType;

  constructor({
    methods,
    inputSchema,
    outputSchema,
    handler,
    resultHandler,
    getOperationId = () => undefined,
    scopes = [],
    middlewares = [],
    tags = [],
    description: long,
    shortDescription: short,
  }: {
    middlewares?: AbstractMiddleware[];
    inputSchema: IN;
    outputSchema: OUT;
    handler: Handler<z.output<IN>, z.input<OUT>, OPT>;
    resultHandler: AbstractResultHandler;
    description?: string;
    shortDescription?: string;
    getOperationId?: (method: Method) => string | undefined;
    methods?: Method[];
    scopes?: string[];
    tags?: string[];
  }) {
    super();
    this.#handler = handler;
    this.#resultHandler = resultHandler;
    this.#middlewares = middlewares;
    this.#getOperationId = getOperationId;
    this.#methods = Object.freeze(methods);
    this.#scopes = Object.freeze(scopes);
    this.#tags = Object.freeze(tags);
    this.#descriptions = { long, short };
    this.#schemas = { input: inputSchema, output: outputSchema };
    this.#responses = {
      positive: Object.freeze(resultHandler.getPositiveResponse(outputSchema)),
      negative: Object.freeze(resultHandler.getNegativeResponse()),
    };
    this.#requestType = hasUpload(inputSchema)
      ? "upload"
      : hasRaw(inputSchema)
        ? "raw"
        : "json";
  }

  public override clone() {
    return new Endpoint({
      // @todo internal props may be unfrozen, but methods returning them can freeze
      methods: this.#methods?.slice(),
      inputSchema: this.#schemas.input,
      outputSchema: this.#schemas.output,
      handler: this.#handler,
      resultHandler: this.#resultHandler,
      getOperationId: this.#getOperationId,
      scopes: this.#scopes.slice(),
      middlewares: this.#middlewares,
      tags: this.#tags.slice(),
      description: this.#descriptions.long,
      shortDescription: this.#descriptions.short,
    }) as this;
  }

  public override getDescription(variant: DescriptionVariant) {
    return this.#descriptions[variant];
  }

  public override getMethods() {
    return this.#methods;
  }

  public override getSchema(variant: "input"): IN;
  public override getSchema(variant: "output"): OUT;
  public override getSchema(variant: IOVariant) {
    return this.#schemas[variant];
  }

  public override getRequestType() {
    return this.#requestType;
  }

  public override getResponses(variant: ResponseVariant) {
    return this.#responses[variant];
  }

  public override getSecurity() {
    return this.#middlewares
      .map((middleware) => middleware.getSecurity())
      .filter((entry) => entry !== undefined);
  }

  public override getScopes() {
    return this.#scopes;
  }

  public override getTags() {
    return this.#tags;
  }

  public override getOperationId(method: Method): string | undefined {
    return this.#getOperationId(method);
  }

  async #parseOutput(output: z.input<OUT>) {
    try {
      return (await this.#schemas.output.parseAsync(output)) as FlatObject;
    } catch (e) {
      throw e instanceof z.ZodError ? new OutputValidationError(e) : e;
    }
  }

  async #runMiddlewares({
    method,
    logger,
    options,
    response,
    ...rest
  }: {
    method: Method | AuxMethod;
    input: Readonly<FlatObject>; // Issue #673: input is immutable, since this.inputSchema is combined with ones of middlewares
    request: Request;
    response: Response;
    logger: ActualLogger;
    options: Partial<OPT>;
  }) {
    for (const mw of this.#middlewares) {
      if (method === "options" && !(mw instanceof ExpressMiddleware)) continue;
      Object.assign(
        options,
        await mw.execute({ ...rest, options, response, logger }),
      );
      if (response.writableEnded) {
        logger.warn(
          "A middleware has closed the stream. Accumulated options:",
          options,
        );
        break;
      }
    }
  }

  async #parseAndRunHandler({
    input,
    ...rest
  }: {
    input: Readonly<FlatObject>;
    options: OPT;
    logger: ActualLogger;
  }) {
    let finalInput: z.output<IN>; // final input types transformations for handler
    try {
      finalInput = (await this.#schemas.input.parseAsync(
        input,
      )) as z.output<IN>;
    } catch (e) {
      throw e instanceof z.ZodError ? new InputValidationError(e) : e;
    }
    return this.#handler({ ...rest, input: finalInput });
  }

  async #handleResult({
    error,
    ...rest
  }: {
    error: Error | null;
    request: Request;
    response: Response;
    logger: ActualLogger;
    input: FlatObject;
    output: FlatObject | null;
    options: Partial<OPT>;
  }) {
    try {
      await this.#resultHandler.execute({ ...rest, error });
    } catch (e) {
      lastResortHandler({
        ...rest,
        error: new ResultHandlerError(ensureError(e), error || undefined),
      });
    }
  }

  public override async execute({
    request,
    response,
    logger,
    config,
  }: {
    request: Request;
    response: Response;
    logger: ActualLogger;
    config: CommonConfig;
  }) {
    const method = getActualMethod(request);
    const options: Partial<OPT> = {};
    let output: FlatObject | null = null;
    let error: Error | null = null;
    const input = getInput(request, config.inputSources);
    try {
      await this.#runMiddlewares({
        method,
        input,
        request,
        response,
        logger,
        options,
      });
      if (response.writableEnded) return;
      if (method === "options") return void response.status(200).end();
      output = await this.#parseOutput(
        await this.#parseAndRunHandler({
          input,
          logger,
          options: options as OPT, // ensured the complete OPT by writableEnded condition and try-catch
        }),
      );
    } catch (e) {
      error = ensureError(e);
    }
    await this.#handleResult({
      input,
      output,
      request,
      response,
      error,
      logger,
      options,
    });
  }
}
