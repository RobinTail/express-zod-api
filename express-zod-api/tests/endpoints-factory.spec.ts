import { RequestHandler } from "express";
import createHttpError from "http-errors";
import { expectTypeOf } from "vitest";
import {
  EndpointsFactory,
  Middleware,
  defaultEndpointsFactory,
  ResultHandler,
  testMiddleware,
} from "../src";
import { EmptyObject } from "../src/common-helpers";
import { Endpoint } from "../src/endpoint";
import { z } from "zod/v4";

describe("EndpointsFactory", () => {
  const resultHandlerMock = new ResultHandler({
    positive: z.string(),
    negative: z.string(),
    handler: vi.fn(),
  });

  describe(".constructor()", () => {
    test("Should create the empty factory with result handler", () => {
      const factory = new EndpointsFactory(resultHandlerMock);
      expect(factory).toBeInstanceOf(EndpointsFactory);
      expect(factory["middlewares"]).toStrictEqual([]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });

    test("Should create the factory with middleware and result handler", () => {
      const middleware = new Middleware({
        input: z.object({ n: z.number() }),
        handler: vi.fn<any>(),
      });
      const factory = new EndpointsFactory(resultHandlerMock).addMiddleware(
        middleware,
      );
      expect(factory["middlewares"]).toStrictEqual([middleware]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });
  });

  describe(".addMiddleware()", () => {
    test("Should create a new factory with a middleware and the same result handler", () => {
      const factory = new EndpointsFactory(resultHandlerMock);
      const middleware = new Middleware({
        input: z.object({ n: z.number() }),
        handler: vi.fn<any>(),
      });
      const newFactory = factory.addMiddleware(middleware);
      expect(factory["middlewares"]).toStrictEqual([]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
      expect(newFactory["middlewares"]).toStrictEqual([middleware]);
      expect(newFactory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });

    test("Should maintain the chain of options", () => {
      defaultEndpointsFactory
        .addMiddleware(
          new Middleware({
            handler: async () => ({ test: "fist option" }),
          }),
        )
        .addMiddleware(
          new Middleware({
            handler: async ({ options }) => {
              expectTypeOf(options.test).toEqualTypeOf<string>();
              return { second: `another option, ${options.test}` };
            },
          }),
        );
    });

    test("Should accept creation props without input schema", () => {
      const factory = defaultEndpointsFactory.addMiddleware({
        handler: async () => ({ test: "fist option" }),
      });
      expectTypeOf(factory).toEqualTypeOf<
        EndpointsFactory<undefined, EmptyObject & { test: string }>
      >();
    });

    test("Issue #2760: should strip excessive props by default", () => {
      defaultEndpointsFactory.build({
        input: z.object({ foo: z.string() }),
        output: z.object({ foo: z.string() }),
        handler: async ({ input }) => {
          expectTypeOf(input).not.toHaveProperty("bar");
          return input;
        },
      });
    });

    test("Issue #2760: should allow excessive props when using loose object schema", () => {
      defaultEndpointsFactory.build({
        input: z.looseObject({ foo: z.string() }),
        output: z.object({ foo: z.string() }),
        handler: async ({ input }) => {
          expectTypeOf(input).toHaveProperty("bar").toEqualTypeOf<unknown>();
          return input;
        },
      });
    });
  });

  describe(".addOptions()", () => {
    test("Should create a new factory with an empty-input middleware and the same result handler", async () => {
      const factory = new EndpointsFactory(resultHandlerMock);
      const newFactory = factory.addOptions(async () => ({
        option1: "some value",
        option2: "other value",
      }));
      expectTypeOf(newFactory).toEqualTypeOf<
        EndpointsFactory<
          undefined,
          EmptyObject & { option1: string; option2: string }
        >
      >();
      expect(factory["middlewares"]).toStrictEqual([]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
      expect(newFactory["middlewares"].length).toBe(1);
      expect(newFactory["middlewares"][0].schema).toBeUndefined();
      const { output: options } = await testMiddleware({
        middleware: newFactory["middlewares"][0],
      });
      expect(options).toEqual({
        option1: "some value",
        option2: "other value",
      });
      expect(newFactory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });
  });

  describe.each(["addExpressMiddleware" as const, "use" as const])(
    ".%s()",
    (method) => {
      test("Should create a new factory with a native express middleware wrapper", async () => {
        const factory = new EndpointsFactory(resultHandlerMock);
        const middleware: RequestHandler = vi.fn((req, {}, next) => {
          req.body.test = "Here is the test";
          next();
        });
        const newFactory = factory[method](middleware, {
          provider: (req) => ({ result: req.body.test }),
        });
        expect(newFactory["middlewares"].length).toBe(1);
        expect(newFactory["middlewares"][0].schema).toBeUndefined();
        const {
          output: options,
          responseMock,
          requestMock,
        } = await testMiddleware({
          middleware: newFactory["middlewares"][0],
        });
        expect(middleware).toHaveBeenCalledTimes(1);
        expect(middleware).toHaveBeenCalledWith(
          requestMock,
          responseMock,
          expect.any(Function),
        );
        expect(requestMock.body).toHaveProperty("test");
        expect(requestMock.body.test).toBe("Here is the test");
        expect(options).toEqual({ result: "Here is the test" });
      });

      test("Should handle rejects from async middlewares", async () => {
        const factory = new EndpointsFactory(resultHandlerMock);
        const middleware: RequestHandler = vi.fn(async () =>
          assert.fail("Rejected"),
        );
        const newFactory = factory[method](middleware);
        const { responseMock } = await testMiddleware({
          middleware: newFactory["middlewares"][0],
        });
        expect(responseMock._getStatusCode()).toBe(500);
        expect(responseMock._getJSONData()).toEqual({
          error: { message: "Rejected" },
          status: "error",
        });
        expect(middleware).toHaveBeenCalledTimes(1);
      });

      test("Should operate without options provider", async () => {
        const factory = new EndpointsFactory(resultHandlerMock);
        const middleware: RequestHandler = vi.fn((req, {}, next) => {
          req.body.test = "Here is the test";
          next();
        });
        const newFactory = factory[method](middleware);
        expect(newFactory["middlewares"].length).toBe(1);
        const {
          output: options,
          responseMock,
          requestMock,
        } = await testMiddleware({
          middleware: newFactory["middlewares"][0],
        });
        expect(middleware).toHaveBeenCalledTimes(1);
        expect(middleware).toHaveBeenCalledWith(
          requestMock,
          responseMock,
          expect.any(Function),
        );
        expect(requestMock.body).toHaveProperty("test");
        expect(requestMock.body.test).toBe("Here is the test");
        expect(options).toEqual({});
      });

      test("Should handle errors", async () => {
        const factory = new EndpointsFactory(resultHandlerMock);
        const middleware: RequestHandler = vi.fn(({}, {}, next) => {
          next(new Error("This one has failed"));
        });
        const newFactory = factory[method](middleware);
        const { responseMock } = await testMiddleware({
          middleware: newFactory["middlewares"][0],
        });
        expect(responseMock._getStatusCode()).toBe(500);
        expect(responseMock._getJSONData()).toEqual({
          error: { message: "This one has failed" },
          status: "error",
        });
        expect(middleware).toHaveBeenCalledTimes(1);
      });

      test("Should transform errors", async () => {
        const factory = new EndpointsFactory(resultHandlerMock);
        const middleware: RequestHandler = vi.fn(({}, {}, next) => {
          next(new Error("This one has failed"));
        });
        const newFactory = factory[method](middleware, {
          transformer: (err) => createHttpError(401, err.message),
        });
        const { responseMock } = await testMiddleware({
          middleware: newFactory["middlewares"][0],
        });
        expect(responseMock._getStatusCode()).toBe(401);
        expect(responseMock._getJSONData()).toEqual({
          error: { message: "This one has failed" },
          status: "error",
        });
        expect(middleware).toHaveBeenCalledTimes(1);
      });
    },
  );

  describe(".build()", () => {
    test("Should create an endpoint with simple middleware", () => {
      const middleware = new Middleware({
        input: z.object({ n: z.number() }),
        handler: vi.fn<any>(),
      });
      const factory = new EndpointsFactory(resultHandlerMock).addMiddleware(
        middleware,
      );
      const handlerMock = vi.fn();
      const endpoint = factory.build({
        input: z.object({ s: z.string() }),
        output: z.object({ b: z.boolean() }),
        handler: handlerMock,
      });
      expect(endpoint).toBeInstanceOf(Endpoint);
      expect(endpoint.methods).toBeUndefined();
      expect(endpoint.inputSchema).toMatchSnapshot();
      expect(endpoint.outputSchema).toMatchSnapshot();
      expectTypeOf(endpoint.inputSchema._zod.output).toEqualTypeOf<
        { n: number } & { s: string }
      >();
    });

    test("Should create an endpoint with refined object middleware", () => {
      const middleware = new Middleware({
        input: z
          .object({
            a: z.number().optional(),
            b: z.string().optional(),
          })
          .refine((props) => Object.keys(props).length, {
            message: "Should be at least one option specified",
          }),
        handler: vi.fn<any>(),
      });
      const factory = new EndpointsFactory(resultHandlerMock).addMiddleware(
        middleware,
      );
      const endpoint = factory.build({
        input: z.object({ i: z.string() }),
        output: z.object({ o: z.boolean() }),
        handler: vi.fn(),
      });
      expect(endpoint.inputSchema).toMatchSnapshot();
      expect(endpoint.outputSchema).toMatchSnapshot();
      expectTypeOf(endpoint.inputSchema._zod.output).toEqualTypeOf<
        { a?: number; b?: string } & { i: string }
      >();
    });

    test("Should create an endpoint with intersection middleware", () => {
      const middleware = new Middleware({
        input: z.object({ n1: z.number() }).and(z.object({ n2: z.number() })),
        handler: vi.fn<any>(),
      });
      const factory = new EndpointsFactory(resultHandlerMock).addMiddleware(
        middleware,
      );
      const handlerMock = vi.fn();
      const endpoint = factory.build({
        input: z.object({ s: z.string() }),
        output: z.object({ b: z.boolean() }),
        handler: handlerMock,
      });
      expect(endpoint).toBeInstanceOf(Endpoint);
      expect(endpoint.methods).toBeUndefined();
      expect(endpoint.inputSchema).toMatchSnapshot();
      expect(endpoint.outputSchema).toMatchSnapshot();
      expectTypeOf(endpoint.inputSchema._zod.output).toEqualTypeOf<
        { n1: number } & { n2: number } & { s: string }
      >();
    });

    test("Should create an endpoint with union middleware", () => {
      const middleware = new Middleware({
        input: z.object({ n1: z.number() }).or(z.object({ n2: z.number() })),
        handler: vi.fn<any>(),
      });
      const factory = new EndpointsFactory(resultHandlerMock).addMiddleware(
        middleware,
      );
      const handlerMock = vi.fn().mockImplementation((params) => ({
        input: params.input,
        b: true,
      }));
      const endpoint = factory.build({
        input: z.object({ s: z.string() }),
        output: z.object({ b: z.boolean() }),
        handler: handlerMock,
      });
      expect(endpoint).toBeInstanceOf(Endpoint);
      expect(endpoint.methods).toBeUndefined();
      expect(endpoint.inputSchema).toMatchSnapshot();
      expect(endpoint.outputSchema).toMatchSnapshot();
      expectTypeOf(endpoint.inputSchema._zod.output).toEqualTypeOf<
        { s: string } & ({ n1: number } | { n2: number })
      >();
    });

    test("should create an endpoint without input schema", () => {
      const factory = new EndpointsFactory(resultHandlerMock);
      const endpoint = factory.build({
        method: "get",
        deprecated: true,
        output: z.object({}),
        handler: vi.fn(),
      });
      expectTypeOf(
        endpoint.inputSchema._zod.output,
      ).toEqualTypeOf<EmptyObject>();
      expect(endpoint.isDeprecated).toBe(true);
    });
  });

  describe(".buildVoid()", () => {
    test("Should be a shorthand for empty object output", () => {
      const factory = new EndpointsFactory(resultHandlerMock);
      const endpoint = factory.buildVoid({
        handler: async () => {},
      });
      expect(endpoint.outputSchema).toMatchSnapshot();
      expectTypeOf(endpoint.outputSchema.shape).toExtend<EmptyObject>();
    });
  });
});
