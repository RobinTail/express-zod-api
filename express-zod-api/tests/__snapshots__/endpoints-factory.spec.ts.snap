// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EndpointsFactory > .build() > Should create an endpoint with intersection middleware 1`] = `
{
  "allOf": [
    {
      "allOf": [
        {
          "properties": {
            "n1": {
              "type": "number",
            },
          },
          "required": [
            "n1",
          ],
          "type": "object",
        },
        {
          "properties": {
            "n2": {
              "type": "number",
            },
          },
          "required": [
            "n2",
          ],
          "type": "object",
        },
      ],
    },
    {
      "properties": {
        "s": {
          "type": "string",
        },
      },
      "required": [
        "s",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with intersection middleware 2`] = `
{
  "properties": {
    "b": {
      "type": "boolean",
    },
  },
  "required": [
    "b",
  ],
  "type": "object",
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with refined object middleware 1`] = `
{
  "allOf": [
    {
      "properties": {
        "a": {
          "type": "number",
        },
        "b": {
          "type": "string",
        },
      },
      "required": [],
      "type": "object",
    },
    {
      "properties": {
        "i": {
          "type": "string",
        },
      },
      "required": [
        "i",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with refined object middleware 2`] = `
{
  "properties": {
    "o": {
      "type": "boolean",
    },
  },
  "required": [
    "o",
  ],
  "type": "object",
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with simple middleware 1`] = `
{
  "allOf": [
    {
      "properties": {
        "n": {
          "type": "number",
        },
      },
      "required": [
        "n",
      ],
      "type": "object",
    },
    {
      "properties": {
        "s": {
          "type": "string",
        },
      },
      "required": [
        "s",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with simple middleware 2`] = `
{
  "properties": {
    "b": {
      "type": "boolean",
    },
  },
  "required": [
    "b",
  ],
  "type": "object",
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with union middleware 1`] = `
{
  "allOf": [
    {
      "oneOf": [
        {
          "properties": {
            "n1": {
              "type": "number",
            },
          },
          "required": [
            "n1",
          ],
          "type": "object",
        },
        {
          "properties": {
            "n2": {
              "type": "number",
            },
          },
          "required": [
            "n2",
          ],
          "type": "object",
        },
      ],
    },
    {
      "properties": {
        "s": {
          "type": "string",
        },
      },
      "required": [
        "s",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`EndpointsFactory > .build() > Should create an endpoint with union middleware 2`] = `
{
  "properties": {
    "b": {
      "type": "boolean",
    },
  },
  "required": [
    "b",
  ],
  "type": "object",
}
`;

exports[`EndpointsFactory > .buildVoid() > Should be a shorthand for empty object output 1`] = `
{
  "properties": {},
  "required": [],
  "type": "object",
}
`;
